File: design-patterns/go/iterator/pagination/page_list.go
// page_list.go

package main

type PageList struct {
	pages []Page
}

func NewPageList() (pageList *PageList) {
	pageList = &PageList{}
	return
}

func (pageList *PageList) Add(page Page) {
	pageList.pages = append(pageList.pages, page)
}

func (pageList *PageList) Remove(page Page) {
	for i, pageElem := range pageList.pages {
		if pageElem == page {
			pageList.pages = append(pageList.pages[:i], pageList.pages[i+1:]...)
			break;
		}
	}
}

func (pageList *PageList) Iterator() AbstractIterator {
	return NewIterator(pageList.pages)
}

File: design-patterns/ts/iterator/pagination/page-list.ts
// page-list.ts

import AbstractIterator from "./abstract-iterator";
import AbstractPageList from "./abstract-page-list";
import Iterator from "./iterator";
import Page from "./page";


class PageList implements AbstractPageList {
    private pages = new Map<number, Page>();;

    add(page: Page): void {
        this.pages.set(page.getNumber(), page);
    }

    remove(page: Page): void {
        this.pages.delete(page.getNumber());
    }

    iterator(): AbstractIterator {
        return new Iterator(this.pages);
    }
}

export default PageList;